### DISCUSSION TITLE
Dziwne skalowanie horyzontalne w ramach EP3

### DISCUSSION THREAD

**Discussion by kmadof** [AzureCommunityPL/group#35] [Pomoc]
Mam Azure Function postawiony na planie EP3 z takimi ustawieniami w skalowania horyzontalnego:

![image](https://user-images.githubusercontent.com/1680172/91415705-ba5d6f00-e84e-11ea-8a4b-6cfdcae54c94.png)

kiedy jednak wejdę w statystki `Live` na App Insight

 
![image](https://user-images.githubusercontent.com/1680172/91415897-ed9ffe00-e84e-11ea-8041-ae91a8d1c892.png)

to mam tam na liście 10 serwerów, które nie wydają się być obciążone. Czy ktoś spotkał się z takim zachowaniem i jest w stanie powiedzieć mi dlaczego tak się dzieje?

**Comment by kamil-mrzyglod** [comment]
A to nie są hosty zombie? W sensie - widzisz to na świeżo postawionym planie?

**Reply by kmadof** [reply]
Plan nie jest świeży :) Jest jakaś opcja, żeby sprawdzić czy to zombie czy nie?

**Reply by kamil-mrzyglod** [reply]
"Twardej" metody na to nie mam, szczególnie że jak host zrobi scale-out, to on ma tzw. grace period kiedy jest utrzymywany żywy. Po tym czasie scale controller zaczyna ubijać niepotrzebne instancje. Nie pamiętam dokładnego interwału, musiałbym poszukać w kodzie bo nie jestem pewien czy w dokumentacji gdzieś je ta informacja.

**Reply by kmadof** [reply]
Ok. Rozumiem. W tym przypadku to jest tak, że kiedy tam nie zajrzę jest tam ich 10. Nawet jeżeli, przez długi czas nie ma większego ruchu. Dodatkowo wszedłem teraz w `Process Explorer` z poziomu `Azure Functions` i tam jest tylko jedna instancja, ale w tych `Live Metrics` w okienku `Sample Telemetry` pojawiają się wpisy że inne instance (niż tylko ta z `Process Explorer`) obsługują ruch.

**Reply by kamil-mrzyglod** [reply]
To może jeszcze inaczej - jaki masz trigger? Przykładowo jeśli miałbyś triggera na Event Hub na powiedzmy 16 partycji, to generalnie tyle mógłbyś widzieć instancji przez większość czasu. To jest trochę poza Twoim zasięgiem, logika scale controllera jest o ile dobrze pamiętam zamknięta i niedostępna do wglądu. 

Tak na samym końcu - pytasz bardziej z ciekawości czy napotykasz jakieś problemy związane z taką ilością działających hostów? :)

**Reply by kmadof** [reply]
Trigger jest na http. Sama funkcja jest właściwie proxy do Event Huba - pobiera dane z body, lekko je modyfikuje i wrzuca do Event Huba. Z kolei Partition Count Event Huba wynosi 1.

Problemów nie ma póki co. Chciałbym zrozumieć przyczynę ;) No i w sumie kwestia kosztu.

Ponadto te query (mam je [stąd](https://github.com/Azure/azure-functions-python-worker/issues/359)):
```
requests
| where timestamp >= ago(1h)
| distinct cloud_RoleInstance, appName
```

zwraca mi 5 wyników. W każdym razie mega dzięki za pomoc. Odpalę jeszcze stres testy i zobaczę jak ruch się rozłoży.

**Reply by kamil-mrzyglod** [reply]
Przy HTTP jest tak, że alokuje się nową instancję maksymalnie co sekundę jeśli jest potrzeba a dodatkowo masz `maxConcurrentRequests`, który jest multiplikowany per działającego hosta. Tu jest wiele zmiennych, które mogą wpływać na działanie Function App - czas potrzebny na zakończenie żądania, faktyczna konfiguracja itd.

**Reply by kmadof** [reply]
Zrobiłem pewien test. Na środowisku testowym przeskalowałem App Plan do EP3 i zasymulowałem ruch podobny do tego na prodzie. W wyniku czego miałem tylko 2 instancje. Z kolei jak puściłem test obciążeniowy to liczba instancji podskoczyła so 22, ale po teście szybko spadła do 1. Dzielę się tym tak dla potomności.
